framework/SupervisedLearning/SyntheticHistory.py

```diff
@@ -0,0 +1,202 @@
+    specs = InputData.parameterInputFactory('SyntheticHistory', strictMode=True,
+        descr=r"""A ROM for characterizing and generating synthetic histories. This ROM makes use of
+               a variety of TimeSeriesAnalysis (TSA) algorithms to characterize and generate new
+               signals based on training signal sets. """)
+    for typ in TSA.knownTypes():
+      c = TSA.returnClass(typ, None) # TODO no message handler for second argument
+      specs.addSub(c.getInputSpecification())
+    return specs
+
+  ### INHERITED METHODS ###
+  def __init__(self, messageHandler, **kwargs):
+    """
+      A constructor that will appropriately intialize a supervised learning object
+      @ In, messageHandler: a MessageHandler object in charge of raising errors,
+                           and printing messages
+      @ In, kwargs: an arbitrary dictionary of keywords and values
+    """
+    # general infrastructure
+    supervisedLearning.__init__(self, messageHandler, **kwargs)
+    self.printTag = 'SyntheticHistoryROM'
+    self._dynamicHandling = True # This ROM is able to manage the time-series on its own.
+    # training storage
+    self.trainedParams = {} # holds results of training each
+    self.tsaAlgorithms = [] # list and order for tsa algorithms to use
+    # data manipulation
+    self.pivotParameterID = None      # string name for time-like pivot parameter
+    self.pivotParameterValues = None  # In here we store the values of the pivot parameter (e.g. Time)
+
+    inputSpecs = kwargs['paramInput']
+    self.readInputSpecs(inputSpecs)
+
+  def readInputSpecs(self, inp):
+    """
+      Reads in the inputs from the user
+      @ In, inp, InputData.InputParams, input specifications
+      @ Out, None
+    """
+    self.pivotParameterID = inp.findFirst('pivotParameter').value # TODO does a base class do this?
+    for sub in inp.subparts:
+      if sub.name in TSA.knownTypes():
+        algo = TSA.returnInstance(sub.name, self.messageHandler)
+        algo.handleInput(sub)
+        self.tsaAlgorithms.append(algo)
+    if self.pivotParameterID not in self.target:
+      self.raiseAnError(IOError, 'The pivotParameter must be included in the target space.')
+
+  def __trainLocal__(self, featureVals, targetVals):
+    """
+      Perform training on input database stored in featureVals.
+      @ In, featureVals, array, shape=[n_timeStep, n_dimensions], an array of input data # Not use for ARMA training
+      @ In, targetVals, array, shape = [n_timeStep, n_dimensions], an array of time series data
+      @ Out, None
+    """
+    self.raiseADebug('Training...')
+    # TODO obtain pivot parameter
+    pivotName = self.pivotParameterID
+    pivotIndex = self.target.index(pivotName)
+    # TODO assumption: only one training signal
+    pivots = targetVals[0, :, pivotIndex]
+    self.pivotParameterValues = pivots[:] # TODO any way to avoid storing these?
+    residual = targetVals[:, :, :] # deep-ish copy, so we don't mod originals
```

Comment:
Would it be more clear to use the `deepcopy` method for this?

Reply:
I'm not sure what the memory difference is on that. I can check here momentarily.

Reply:
on consideration, I like the clarity of the way it is in terms of dimensionality; it could be inferred elsewhere, but since this is the first time "residual" has come into existence, it provides some hints to its structure.

Further, a deep copy would blissfully copy any structure that's there; this in some way assures there's the expected three-dimensional numpy array structure.

Reply:
> this in some way assures there's the expected three-dimensional numpy array structure.

This is a good point. I especially like when I can see the structure of objects in python. Personally, I would love to utilize python's type-hints for things like this.